# Import Libraries

In [60]:
import numpy as np
import pandas as pd
pd.set_option("max_columns", None)
from scipy.stats import chi2_contingency

# Import Data

In [86]:
data = pd.read_csv('fashion_data_on_Instagram.csv')
data.set_index('UserId', inplace = True)
data.head()

Followings  Followers  MediaCount  \
UserId                                                              
1171579752855683619_212070047        518.0     9840.0         541   
1171594777274371222_176762322       7333.0     2300.0         272   
1171407552643586413_581125501        131.0      605.0         106   
1171407552643586413_581125501        131.0      605.0         106   
1171508187966229230_2797323089       132.0      456.0          56   

                                  BrandName BrandCategory  \
UserId                                                      
1171579752855683619_212070047   abercrombie   High street   
1171594777274371222_176762322   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171508187966229230_2797323089  abercrombie   High street   

                                                                         Hashtags  \
UserId                                                                              
1171579752855683619_212070047   beautiful, summer, fashion, love, cute, food, ...   
1171594777274371222_176762322   teen, model, brunette, selfie, hollister, snap...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171508187966229230_2797323089  abercrombieandfitch, shopping, love, hollister...   

                                                                          Caption  \
UserId                                                                              
1171579752855683619_212070047     We were born to be REAL, not to be perfect. ...   
1171594777274371222_176762322     Gotta run, but first, let me take selfie. Me...   
1171407552643586413_581125501                         dress to impress\U0001f457    
1171407552643586413_581125501                         dress to impress\U0001f457    
1171508187966229230_2797323089    #love #shopping #shoppen #hollister #abercro...   

                                                                           ImgURL  \
UserId                                                                              
1171579752855683619_212070047   https://scontent.cdninstagram.com/t51.2885-15/...   
1171594777274371222_176762322   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171508187966229230_2797323089  https://scontent.cdninstagram.com/t51.2885-15/...   

                                Likes  Comments   CreationTime  \
UserId                                                           
1171579752855683619_212070047      97          0    1453883211   
1171594777274371222_176762322      94          0    1453885002   
1171407552643586413_581125501      91          3    1453862683   
1171407552643586413_581125501      94          3    1453862683   
1171508187966229230_2797323089      9          1    1453874680   

                                                                    Link  \
UserId                                                                     
1171579752855683619_212070047   https://www.instagram.com/p/BBCSZPzAMIj/   
1171594777274371222_176762322   https://www.instagram.com/p/BBCVz4YJEyW/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171508187966229230_2797323089  https://www.instagram.com/p/BBCCH1zMEru/   

                                  Selfie  BodySnap  Marketing  ProductOnly  \
UserId                                                                       
1171579752855683619_212070047   0.000885  0.336438   0.741220     0.009454   
1171594777274371222_176762322   0.003912  0.995503   0.001728     0.002125   
1171407552643586413_5811

# Clean Data

### Let's See if there are any NaNs in the data

In [17]:
set(data['Hashtags'].isna())

{False, True}

In [18]:
set(data['Caption'].isna())

{False, True}

In [50]:
set(data['Comments '].isna())

{False}

##### So there exist NaN values in Hashtags and Captions. There are no other columns that contain NaN.

In [55]:
data.isna().groupby('Hashtags')['Followings'].count()

Hashtags
False    24606
True       146
Name: Followings, dtype: int64

##### There are 146 NaN rows in hashtags. It is possible that this is simply due to the fact that some posts do not have hashtags.

In [57]:
data.isna().groupby('Caption')['Followings'].count()

Caption
False    24717
True        35
Name: Followings, dtype: int64

##### There are 35 NaN rows in Caption

### Maybe the caption or hashtag information is missing because there is either only a caption or only hashtags....

In [69]:
hashtag_nan = list(data['Hashtags'].isna())
caption_nan = list(data['Caption'].isna())

In [73]:
count = 0
for i in range(len(hashtag_nan)):
    if hashtag_nan[i] and caption_nan[i]:
        count += 1
print('The number of times both hashtag_nan and caption_nan is true is', count)

The number of times both hashtag_nan and caption_nan is true is 18


##### So we have missing hashtag both 18/35 times that we have a missing caption. So it is not exactly correlated. Maybe we can just drop the rows that have missing caption or hashtag if we need to use these, or claim that there was not hashtag or caption for these rows.

### I want to try and see if I can make brand category numeric

In [87]:
brand_category = pd.get_dummies(data['BrandCategory'])

In [88]:
brand_category.head()

Designer  High street  Mega couture  \
UserId                                                                
1171579752855683619_212070047          0            1             0   
1171594777274371222_176762322          0            1             0   
1171407552643586413_581125501          0            1             0   
1171407552643586413_581125501          0            1             0   
1171508187966229230_2797323089         0            1             0   

                                Small couture  
UserId                                         
1171579752855683619_212070047               0  
1171594777274371222_176762322               0  
1171407552643586413_581125501               0  
1171407552643586413_581125501               0  
1171508187966229230_2797323089              0

In [94]:
result = pd.concat([data, brand_category], sort = False, axis = 1)
result.head()

Followings  Followers  MediaCount  \
UserId                                                              
1171579752855683619_212070047        518.0     9840.0         541   
1171594777274371222_176762322       7333.0     2300.0         272   
1171407552643586413_581125501        131.0      605.0         106   
1171407552643586413_581125501        131.0      605.0         106   
1171508187966229230_2797323089       132.0      456.0          56   

                                  BrandName BrandCategory  \
UserId                                                      
1171579752855683619_212070047   abercrombie   High street   
1171594777274371222_176762322   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171407552643586413_581125501   abercrombie   High street   
1171508187966229230_2797323089  abercrombie   High street   

                                                                         Hashtags  \
UserId                                                                              
1171579752855683619_212070047   beautiful, summer, fashion, love, cute, food, ...   
1171594777274371222_176762322   teen, model, brunette, selfie, hollister, snap...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171407552643586413_581125501   alexandani, cute, llbean, beanboots, anthropol...   
1171508187966229230_2797323089  abercrombieandfitch, shopping, love, hollister...   

                                                                          Caption  \
UserId                                                                              
1171579752855683619_212070047     We were born to be REAL, not to be perfect. ...   
1171594777274371222_176762322     Gotta run, but first, let me take selfie. Me...   
1171407552643586413_581125501                         dress to impress\U0001f457    
1171407552643586413_581125501                         dress to impress\U0001f457    
1171508187966229230_2797323089    #love #shopping #shoppen #hollister #abercro...   

                                                                           ImgURL  \
UserId                                                                              
1171579752855683619_212070047   https://scontent.cdninstagram.com/t51.2885-15/...   
1171594777274371222_176762322   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171407552643586413_581125501   https://scontent.cdninstagram.com/t51.2885-15/...   
1171508187966229230_2797323089  https://scontent.cdninstagram.com/t51.2885-15/...   

                                Likes  Comments   CreationTime  \
UserId                                                           
1171579752855683619_212070047      97          0    1453883211   
1171594777274371222_176762322      94          0    1453885002   
1171407552643586413_581125501      91          3    1453862683   
1171407552643586413_581125501      94          3    1453862683   
1171508187966229230_2797323089      9          1    1453874680   

                                                                    Link  \
UserId                                                                     
1171579752855683619_212070047   https://www.instagram.com/p/BBCSZPzAMIj/   
1171594777274371222_176762322   https://www.instagram.com/p/BBCVz4YJEyW/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171407552643586413_581125501   https://www.instagram.com/p/BBBrPZ2h9lt/   
1171508187966229230_2797323089  https://www.instagram.com/p/BBCCH1zMEru/   

                                  Selfie  BodySnap  Marketing  ProductOnly  \
UserId                                                                       
1171579752855683619_212070047   0.000885  0.336438   0.741220     0.009454   
1171594777274371222_176762322   0.003912  0.995503   0.001728     0.002125   
1171407552643586413_5811

##### Now the brand category is represented in the last four columns of the dataframe with one_hot_encoding.

# Model with Numeric Data Only

First, lets try to build a model with only the numeric data. This will be easy to do and will allow us to understand whether any of these numeric factors influence the number of likes/comments a user gets. 